File: backend/app/utils/event_status.py
from datetime import datetime
from app.models import Event


ACTIVE_EVENT_STATUSES = ["ACTIVE", "UPCOMING", "ONGOING"]
BLOCKED_EVENT_STATUSES = ["CANCELLED", "COMPLETED"]


def is_event_expired(event: Event, now: datetime | None = None) -> bool:
    if not event or not event.event_date:
        return False

    current_time = now or datetime.now()
    return event.status == "COMPLETED" or event.event_date < current_time


def is_event_bookable(event: Event, now: datetime | None = None) -> bool:
    if not event:
        return False

    return event.status not in BLOCKED_EVENT_STATUSES and not is_event_expired(
        event,
        now,
    )


def expire_past_events(db):
    now = datetime.now()

    expired_events = (
        db.query(Event)
        .filter(Event.event_date < now)
        .filter(Event.status.in_(ACTIVE_EVENT_STATUSES))
        .all()
    )

    for event in expired_events:
        event.status = "COMPLETED"

    db.commit()

File: frontend/src/utils/eventStatus.js
export function isEventExpired(event) {
  if (!event) return false;
  if (event.status === "COMPLETED") return true;
  if (!event.event_date) return false;

  const eventDate = new Date(event.event_date);
  if (Number.isNaN(eventDate.getTime())) return false;

  return eventDate < new Date();
}

export function isEventCancelled(event) {
  return event?.status === "CANCELLED";
}

export function isEventUnavailable(event) {
  return isEventExpired(event) || isEventCancelled(event);
}

export function getDisplayEventStatus(event) {
  if (isEventCancelled(event)) return "CANCELLED";
  if (isEventExpired(event)) return "EXPIRED";
  return event?.status || "UPCOMING";
}

export function isEventActive(event) {
  return ["ACTIVE", "UPCOMING", "ONGOING"].includes(event?.status) && !isEventExpired(event);
}
